src_lockfree/htbl/htbl.body.ml

```diff
@@ -0,0 +1,617 @@
+
+let create (type k) ?hashed_type ?min_buckets ?max_buckets () =
+  let min_buckets =
+    match min_buckets with
+    | None -> min_buckets_default
+    | Some n ->
+        let n = Int.max lo_buckets n |> Int.min hi_buckets in
+        ceil_pow_2_minus_1 (n - 1) + 1
+  in
+  let max_buckets =
+    match max_buckets with
+    | None -> Int.max min_buckets max_buckets_default
+    | Some n ->
+        let n = Int.max min_buckets n |> Int.min hi_buckets in
+        ceil_pow_2_minus_1 (n - 1) + 1
+  in
+  let equal, hash =
+    match hashed_type with
+    | None ->
+        (( = ), Stdlib.Hashtbl.seeded_hash (Int64.to_int (Random.bits64 ())))
+    | Some ((module Hashed_type) : k hashed_type) ->
+        (Hashed_type.equal, Hashed_type.hash)
+  in
+  {
+    hash;
+    buckets =
+      Atomic_array.make min_buckets
+        (B (Nil_with_size { size_modifier = Size.used_once }));
+    equal;
+    size = Size.create ();
+    pending = Nothing;
+    min_buckets;
+    max_buckets;
+  }
+  |> Atomic.make_contended
+
+(* *)
+
+let hashed_type_of (type k) (t : (k, _) t) : k hashed_type =
+  let r = Atomic.get t in
+  (module struct
+    type t = k
+
+    let hash = r.hash
+    and equal = r.equal
+  end)
+
+let min_buckets_of t = (Atomic.get t).min_buckets
+let max_buckets_of t = (Atomic.get t).max_buckets
+
+(* *)
+
+let rec take_at backoff size bs i =
+  match Atomic_array.unsafe_fenceless_get bs i with
+  | B
+      ((Nil_with_size { size_modifier } | Cons_with_size { size_modifier; _ })
+       as spine) -> begin
+      if size_modifier != Size.used_once then
+        Size.update_once size size_modifier;
+
+      if
+        Atomic_array.unsafe_compare_and_set bs i (B spine)
+          (B (Resize { spine }))
+      then spine
+      else take_at (Backoff.once backoff) size bs i
+    end
+  | B (Resize spine_r) -> spine_r.spine
+
+let rec copy_all r target i t step =
+  let i = (i + step) land (Atomic_array.length target - 1) in
+  let spine = take_at Backoff.default r.size r.buckets i in
+  let (B before) = Atomic_array.unsafe_fenceless_get target i in
+  (* The [before] value is physically different for each resize and so checking
+     that the resize has not finished is sufficient to ensure that the
+     [compare_and_set] below does not disrupt the next resize. *)
+  Atomic.get t == r
+  && begin
+       begin
+         match before with
+         | Resize _ ->
+             Atomic_array.unsafe_compare_and_set target i (B before) (B spine)
+             |> ignore
+         | Nil_with_size _ | Cons_with_size _ -> ()
+       end;
+       i = 0 || copy_all r target i t step
+     end
+
+(* *)
+
+let[@tail_mod_cons] rec filter_ t msk chk = function
+  | Nil -> Nil
+  | Cons r ->
+      if t r.key land msk = chk then
+        Cons { r with rest = filter_ t msk chk r.rest }
+      else filter_ t msk chk r.rest
+
+let[@tail_mod_cons] rec filter_fst t mask chk = function
+  | Nil -> Nil_with_size { size_modifier = Size.used_once }
```

Comment:
`Nil_with_size { size_modifier = Size.used_once }` is logically a constant value.  However, because the `size_modifier` field is `mutable`, the compiler cannot optimize this to a constant value.  This means that a new value is allocated every time.  There are a couple of other places in the code where a similar `Nil_with_size { size_modifier = Size.used_once }` allocation happens unnecessarily.  Avoiding the allocations could slightly reduce memory use and improve performance.

Reply:
(took me a minute to get it, but got it, thanks !)

Reply:
So I defined a `nil_with_used_once` value to used in this case but it causes type issue with weak type variable. No sure if this is what you wanted me to do.

Reply:
There are a few ways to avoid this problem:
* You could allow `Nil` at the root level of a bucket.
* You could include a `nil_with_used_once` as a field in the hash table record.
* You could use an unsafe cast to have a single `nil_with_used_once` value at the module level.